pkg/liquidity-mining/contracts/fee-distribution/FeeDistributor.sol

```diff
@@ -0,0 +1,501 @@
+                userEpoch += 1;
+                oldUserPoint = userPoint;
+                if (userEpoch > maxUserEpoch) {
+                    userPoint = IVotingEscrow.Point(0, 0, 0, 0);
+                } else {
+                    userPoint = _votingEscrow.user_point_history(user, userEpoch);
+                }
+            } else {
+                // The week being considered lies inside the user epoch described by `oldUserPoint`
+                // we can then use it to calculate the user's balance at the beginning of the week.
+
+                int128 dt = int128(weekCursor - oldUserPoint.ts);
+                uint256 userBalance = oldUserPoint.bias > oldUserPoint.slope * dt
+                    ? uint256(oldUserPoint.bias - oldUserPoint.slope * dt)
+                    : 0;
+
+                // User's lock has expired and they haven't relocked yet.
+                if (userBalance == 0 && userEpoch > maxUserEpoch) break;
+
+                // User had a nonzero lock and so is eligible to collect fees.
+                if (userBalance > 0) {
+                    _userBalanceAtTimestamp[user][weekCursor] = userBalance;
+                }
+
+                weekCursor += 1 weeks;
+            }
+        }
+
+        _userLastEpochCheckpointed[user] = userEpoch - 1;
+        _userTimeCursor[user] = weekCursor;
+    }
+
+    /**
+     * @dev Cache the totalSupply of VotingEscrow token at the beginning of each new week
+     */
+    function _checkpointTotalSupply() internal {
+        uint256 timeCursor = _timeCursor;
+        uint256 weekStart = _roundDownTimestamp(block.timestamp);
+
+        // We expect `timeCursor == weekStart + 1 weeks` when fully up to date.
+        if (timeCursor > weekStart) {
+            // We've already checkpointed up to this week so perform early return
+            return;
+        }
+
+        _votingEscrow.checkpoint();
+
+        // Step through the each week and cache the total supply at beginning of week on this contract
+        for (uint256 i = 0; i < 20; ++i) {
+            if (timeCursor > weekStart) break;
+
+            uint256 supplySnapshot = _votingEscrow.totalSupply(timeCursor);
+            if (supplySnapshot > 0) {
+                _veSupplyCache[timeCursor] = supplySnapshot;
+            }
+
+            timeCursor += 1 weeks;
+        }
+        // Update state to the end of the current week (`weekStart` + 1 weeks)
+        _timeCursor = timeCursor;
+    }
+
+    // Helper functions
+
+    /**
+     * @dev Wrapper around `_userTokenTimeCursor` which returns the start timestamp for `token`
+     * if `user` has not attempted to interact with it previously.
+     */
+    function _getUserTokenTimeCursor(address user, IERC20 token) internal view returns (uint256) {
+        uint256 userTimeCursor = _userTokenTimeCursor[user][token];
+        return userTimeCursor > 0 ? userTimeCursor : _tokenStartTime[token];
```

Comment:
Improvement for a later PR, we should also have a user-level start time so that someone who locks veBAL in a year doesn't have to claim 52 empty weeks first.